### PR TITLE
add `attention_mask` to optimizer json

### PR DESCRIPTION
**Pull request by damian0815**
you can set it to "none" (current behaviour, default), "pad" (mask out padding tokens), or "pad_and_bos_eos" (mask out padding and beginning of sequence/end of sequence markers)

leaving at draft state for now, because i would like to get more eyes on it and also i haven't updated the docs

### CODE REVIEW

**Comment by victorchall** [comment]
https://drive.google.com/drive/folders/1lncdsJomPKE-ULrZbocBccw4kAFf_BAH
https://huggingface.co/panopstor/attn_mask_tests
image
image

**Comment by victorchall** [comment]
I think this is dead as it proved not to work, or drastically reduced training speed.  Would need meaningful trainer settings to try again.  Feel free to reopen if you know settings that will work that I could try.
